### PR TITLE
Ncl 5295

### PR DESCRIPTION
**Pull request by janinko**
### Checklist:

* [ ] Have you added unit tests for your change?

### CODE REVIEW

File: rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java [hunk 1]
```diff
@@ -163,4 +163,18 @@ Artifact getSpecific(
...
+    @Operation(summary = "Gets the milestones that produced or consumed this artifact.",
+            responses = {
+                @ApiResponse(responseCode = SUCCESS_CODE, description = SUCCESS_DESCRIPTION,
+                    content = @Content(schema = @Schema(implementation = SwaggerPages.MilestoneInfoPage.class))),
+                @ApiResponse(responseCode = NOT_FOUND_CODE, description = NOT_FOUND_DESCRIPTION),
+                @ApiResponse(responseCode = SERVER_ERROR_CODE, description = SERVER_ERROR_DESCRIPTION,
+                    content = @Content(schema = @Schema(implementation = ErrorResponse.class)))
+    })
+    @GET
+    @Path("/{id}/milestones")
```
Comment (michalovjan) [inline rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java]: what about "/{id}/info" ?

File: dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * This is entry describing milestone that produced or consumed an artifact.
+ *
+ * @author Honza Brázdil &lt;jbrazdil@redhat.com&gt;
+ */
+@Data
+@Builder(builderClassName = "Builder")
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonDeserialize(builder = MilestoneInfo.Builder.class)
+public class MilestoneInfo {
```
Comment (michalovjan) [inline dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java]: Hmm... The name seems is kinda off. You provide much more than Milestone info

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java [hunk 3]
```diff
@@ -184,4 +214,96 @@ public void cancelMilestoneCloseProcess(String id)
 
         return queryForCollection(pageIndex, pageSize, sortingRsql, query, withProductVersionId(Integer.valueOf(productVersionId)));
     }
+
+    @Override
+    public Page<MilestoneInfo> getMilestonesOfArtifact(String id, int pageIndex, int pageSize) {
+        CriteriaBuilder cb = em.getCriteriaBuilder();
+
+        List<Integer> buildIn = getBuildIds(cb, id, Artifact_.buildRecords);
```
Comment (michalovjan) [inline facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java]: You are getting milestone IDs and not IDs of the BuildRecords, the name of a variable is misleading.

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java [hunk 4]
```diff
@@ -184,4 +214,96 @@ public void cancelMilestoneCloseProcess(String id)
...
+        List<Integer> dependencyOf = getBuildIds(cb, id, Artifact_.dependantBuildRecords);
```
Comment (michalovjan) [inline facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java]: Is only one query to DB possible?

File: dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    private final String productVersionId;
+    private final String productVersionVersion;
+    private final String milestoneId;
+    private final String milestoneVersion;
+    private final Instant milestoneEndDate;
+    private final String releaseId;
+    private final String releaseVersion;
+    private final Instant releaseReleaseDate;
+    /**
+     * Wheather the queried artifact was build in this milestone or not.
```
Comment (michalovjan) [inline dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java]: "whether", "built"

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java [hunk 6]
```diff
@@ -184,4 +214,96 @@ public void cancelMilestoneCloseProcess(String id)
...
+        milestoneIds.addAll(buildIn);
+        milestoneIds.remove(null); // some builds are not in a milestone and so it gives us null
+        if(milestoneIds.isEmpty()){
+            return new Page<>();
+        }
+
+        CriteriaQuery<Tuple> query = milestoneInfoQuery(cb, milestoneIds);
+        int offset = pageIndex * pageSize;
+        List<MilestoneInfo> milestones = em.createQuery(query)
+                .setMaxResults(pageSize)
```
Comment (michalovjan) [inline facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java]: Is it possible to use LIMIT in DB query?

Comment (janinko) [reply facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java]: setMaxResults is not setting LIMIT in the DB query?

File: rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java (hunk 1 above)
Comment (janinko) [reply rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java]: @alexcreasy WDYT?

File: dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java (hunk 2 above)
Comment (janinko) [reply dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java]: Well. it provides info about a milestone: In what product and version the milestone is, what is the milestone version, if the milestone was released or not...

File: rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java (hunk 1 above)
Comment (alexcreasy) [reply rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java]: I prefer {id}/milestones myself, {id}/info is too general and this is a very specific query... To me artifact/{id} is artifact "info".

It doesn't really provide more than milestones if you think about the query in userspace and not in db / implementation space.  A milestone, in reality, is a product + product version + milestone.

Person A: "Hey should we upgrade to 1.GA"? 
Person B: "Eh?"
Person A: "I mean 6.0.1.GA"
Person B: "errrrrrrm, you're talking about FUSE, right?"
Person A: "yeah, exactly. FUSE 6.0.1.GA"
Person B: "Right, sure, let's do it..."

;)

Comment (alexcreasy) [reply rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java]: If you would prefer something more generic the first thing that comes to my mind is "{id}/usages"

File: dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java (hunk 2 above)
Comment (alexcreasy) [reply dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java]: +1 with Honza, as above a milestone is really product + ver + milestone , it's decomposed purely for efficiency in computer world IMO.

File: rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java (hunk 1 above)
Comment (michalovjan) [reply rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/ArtifactEndpoint.java]: "/{id}/milestones" is absolutely fine then, didn't know the information was gotten directly from milestone

File: dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java (hunk 2 above)
Comment (michalovjan) [reply dto/src/main/java/org/jboss/pnc/dto/response/MilestoneInfo.java]: Yeah, you are right

**Comment by janinko** [comment]
retest this please

**Comment by janinko** [comment]
retest this please

**Pull request merged**